README.md
# GT Bench

This benchmark evaluates the performance of [GridTools](https://github.com/gridtools/gridtools), an open source C++-embedded DSL for weather and climate codes.
The benchmark implements an advection-diffusion solver using the finite difference method. The benchmark uses operator splitting for temporal integration typical of weather and climate codes, whereby the two horizontal dimensions use an explicit discretization, and the third, vertical, dimension is implicit.

## Dependencies

### GridTools

The benchmark requires version 1.1.2 of GridTools, which can be obtained from the GridTools repository:

```console
$ git clone --branch v1.1.2 https://github.com/GridTools/gridtools.git
```

GridTools further depends on [Boost](https://www.boost.org/) (minimum version 1.67.0). A recent version of [CMake](https://cmake.org/) is required to build and install GridTools (minimum version 3.14.5).

Follow the [GridTools documentation](https://gridtools.github.io/gridtools/latest/index.html) for an installation guideline. Note that GPU support has to be enabled when building GridTools if the benchmark is to be run on a GPU system.

### GHEX

The benchmark requires the GHEX library. It can be obtained using:

```console
$ git clone https://github.com/GridTools/GHEX.git
```

#### Dependencies
Required:
- GridTools (see instructions above)
- [Boost](https://www.boost.org/)
- MPI (for example [OpenMPI](https://github.com/open-mpi/ompi))

Optional:
- [Unified Communication X (UCX)](https://github.com/openucx/ucx)
- [PMIx](https://github.com/openpmix/openpmix)

Additionally, [CMake](https://cmake.org/) is required for building GHEX.

##### Building GHEX

Once all necessary and optional dependencies have been installed, GHEX can be installed using CMake as follows:
```console
$ cd /PATH/TO/GHEX-SOURCE
$ mkdir build && cd build
$ cmake -DCMAKE_INSTALL_PREFIX=/PATH/TO/GHEX-INSTALLATION \
        -DGridTools_DIR=/PATH/TO/GRIDTOOLS-INSTALLATION/lib/cmake \
        ..
```

To enable UCX support, pass additionally the following flags
```console
        -DGHEX_USE_UCP=ON \
        -DUCP_INCLUDE_DIR=/PATH/TO/UCX-INSTALLATION/include \
        -DUCP_LIBRARY=/PATH/TO/UCX-INSTALLATION/lib/libucp.so
```

If PMIx shall be enabled, follow the above pattern by define additionally
```console
        -DGHEX_USE_PMIx=ON \
        -DPMIX_INCLUDE_DIR=/PATH/TO/PMIX-INSTALLATION/include \
        -DPMIX_LIBRARY=/PATH/TO/PMIX-INSTALLATION/lib/libpmix.so
```

After successful configuration, type
```console
$ make install
```

## Building

Once the dependencies are installed, CMake can be used to configure the build as follows:
```console
$ cd /PATH/TO/GTBENCH-SOURCE
$ mkdir build && cd build
$ cmake ..
```

Depending on the setup, the GridTools and GHEX installation directories might have to be specified. This can be accomplished by passing additional arguments to cmake:
```console
$ cmake -DGridTools_DIR=/PATH/TO/GRIDTOOLS-INSTALLATION \
        -DGHEX_DIR=/PATH/TO/GHEX-INSTALLATION \
        ..
```

### Selecting the GridTools Backend

The GridTools backend specifies which hardware architecture to target.
The backend can be selected by setting the `GTBENCH_BACKEND` option when configuring with CMake:
```console
$ cmake -DGTBENCH_BACKEND=<BACKEND> ..
```
where `<BACKEND>` must be either `x86`, `mc`, or `cuda`. The `x86` and `mc` backends are two different CPU-backends of GridTools. On modern CPUs with large vector width and/or many cores, the `mc` backend might perform significantly better. On CPUs without vectorization or small vector width and limited parallelism, the `x86` backend might perform better. The `cuda` backend currently supports running NVIDIA CUDA-capable GPUs and – despite its name – also AMD HIP-capable GPUs.

### Selecting the GPU Compilation Framework

**Note** This section is only relevant for GPU targets.

There are three GPU targets available, which are set at when configuring GridTools by setting the CMake `GT_CUDA_COMPILATION_TYPE` parameter:

1. `NVCC-CUDA`: NVIDIA CUDA compilation using the NVIDIA compiler.
2. `Clang-CUDA`: Clang CUDA compilation using the compiler.
3. `HIPCC-AMDGPU` AMD HIP compilation using AMD’s HIP-Clang compiler. **Note**: the deprecated *HCC* compiler is not supported.


### Selecting the Communication Backend

The benchmark implementation brings several backends for distributed parallelism. These can be selected using the CMake variable `GTBENCH_COMMUNICATION_BACKEND`:
```console
$ cmake -DGTBENCH_COMMUNICATION_BACKEND=<COMM_BACKEND> ..
```
where `COMM_BACKEND` can be `ghex_comm`, `gcl`, `simple_mpi`, `single_node`. `simple_mpi` and `single_node` are for debugging purposes only.
- The `single_node` options is useful for performing "single-node" tests to understand kernel performance.
- The `simple_mpi` implementation uses a simple MPI 2 sided communication for halo exchanges.
- The `gcl` implementation uses a optimized MPI based communication library shipped with [GridTools](https://gridtools.github.io/gridtools/latest/user_manual/user_manual.html#halo-exchanges).
- The `ghex_comm` option will use highly optimized distributed communication via the GHEX library, designed for best performance at scale.
 Additionally, this option will enable a multi-threaded version of the benchmark, where a rank may have more than one sub-domain (over-subscription), which are delegated to separate threads. **Note:** The gridtools computations use openmp threads on the CPU targets which will not be affected by this parameter.

#### Selecting the Transport Layer

If the `ghex_comm` communication backend has been selected, the underlying transport layer will be either 
*UCX* or *MPI*. The behaviour can be chosen by defining the the CMake boolean variables `GHEX_USE_UCP` when configuring the GHEX library, see above.

## Running the Benchmark

### Benchmark

The benchmark executable takes a single command line parameter, the global horizontal domain size `N`. The simulation will then be performed on a total domain size of `N×N×60` grid points. To launch the benchmark use the appropriate MPI launcher (`mpirun`, `mpiexec`, `srun` or similar):
```console
$ mpi_launcher <LAUNCHER_OPTIONS> ./benchmark <N>
```

Example output of a single-node benchmark run:
```
Running GTBENCH
Domain size:             100x100x60
Floating-point type:     float
GridTools backend:       cuda
Communication backend:   single_node
Median time:             0.198082s (95% confidence: 0.19754s - 0.200368s)
Columns per second:      50484.1 (95% confidence: 49908.1 - 50622.6)
```

For testing, the number of runs (and thus the run time) can be reduced as follows:
```console
$ mpi_launcher <LAUNCHER_OPTIONS> ./benchmark <N> <RUNS>
```
For example, run only once:
```console
$ mpi_launcher ./benchmark 24000 1
Running GTBENCH
Domain size:             24000x24000x60
Floating-point type:     float
GridTools backend:       cuda
Communication backend:   ghex_comm
Median time:             8.97857s
Columns per second:      6.41528e+07
```
Note that no confidence intervals are given in this case, but they are required for the final benchmark runs.

Provided the `ghex_comm` communication backend was selected, the executable of the benchmark will accept an additional 
command line option:
```console
$ mpi_launcher <LAUNCHER_OPTIONS> ./benchmark <N> <RUNS> <NUM_THREADS>
```
This will determine the number of threads (and, thus, sub-domains) per rank.


### Convergence Tests

To make sure that the solver converges to the analytical solution of the advection-diffusion equation, we provide convergence tests. They might be helpful for evaluating correctness after possible code changes in the computation or communication backends, or the compiler optimization level. To run them, use:
```console
$ mpi_launcher <LAUNCHER_OPTIONS> ./convergence_tests
```

The convergence tests can be run on 1, 2 or 4 MPI ranks.

Example outputs for single and double precision configurations can be found in the files *convergence_float.out* and *convergence_double.out*.

Note that the expected convergence orders of some tests do not exactly match the theoretical order. This is either due to limited numerical precision or to suboptimal range of tested spatial or temporal resolutions for a specific discretization.


CHANGELOG.md
# Changelog

# February 28, 2020
- updated gtbench source code, commit [4994496ab2e1a62566d10a69eed6a2b3282d4d4c](https://github.com/GridTools/gtbench/commit/4994496ab2e1a62566d10a69eed6a2b3282d4d4c) 
  - repository was moved to [GridTools Github organization](https://github.com/GridTools)
  - license was updated
  - added new `gcl` communication backend
  - updated README accordingly
  - excluded first time step from measurements
  - use different stream synchronization and compilation flags for `cuda` backend
  - added reference performance results
- updated [ghex source code](https://github.com/GridTools/GHEX/tree/lumi), commit [e0f78983b332661dfa8f43d1f71f814b80e44b0d](https://github.com/GridTools/GHEX/commit/e0f78983b332661dfa8f43d1f71f814b80e44b0d)

# initial publication of gtbench


benchmark.cpp
/*
 * gtbench
 *
 * Copyright (c) 2014-2020, ETH Zurich
 * All rights reserved.
 *
 * Please, refer to the LICENSE file in the root directory.
 * SPDX-License-Identifier: BSD-3-Clause
 */
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <thread>
#include <vector>

#include "./communication/backends.hpp"
#include "./execution/run.hpp"
#include "./numerics/solver.hpp"

int main(int argc, char **argv) {
  // user input handling
  std::size_t runs = 101;
  const std::size_t nz = 60;
  int num_threads = 1;

  if (argc < 2 || argc > 4) {
    std::cerr << "Usage: " << argv[0] << " N [RUNS] [THREADS]" << std::endl;
    std::cerr << "N: global domain size, full domain will be NxNx" << nz
              << std::endl;
    std::cerr << "RUNS: number of runs, reported is the median and "
              << "95% confidence intervals (for RUNS > 100, "
              << "default value: " << runs << ")\n"
              << "THREADS: number of threads (sub-domains) per rank, default "
                 "value: 1\n"
              << "(this option affects the GHEX backend only)" << std::endl;
    return 1;
  }

  const std::size_t n = std::atoll(argv[1]);
  if (argc > 2)
    runs = std::atoll(argv[2]);
  if (argc > 3)
    num_threads = std::atoi(argv[3]);

    // override number of threads for simple backends
#if !defined(GTBENCH_USE_GHEX)
  if (num_threads != 1) {
    std::cerr << "number of threads cannot be larger than 1 for this backend."
              << std::endl;
    return 1;
  }
#endif
  const bool multi_threaded = num_threads > 1;

  // communication setup
  auto comm_world = communication::GTBENCH_COMMUNICATION_BACKEND::world(
      argc, argv, multi_threaded);

  auto fmt = [&]() -> std::ostream & {
    return std::cout << std::endl << std::setw(26) << std::left;
  };

  std::cout << "Running GTBENCH";
  fmt() << "Domain size:" << n << "x" << n << "x" << nz;
  fmt() << "Number of domain threads:" << num_threads;

#ifdef _OPENMP
  // get number of openmp threads
  fmt() << "Number of OpenMP threads:" << omp_get_max_threads();
#endif

#define GTBENCH_STR2(var) #var
#define GTBENCH_STR(var) GTBENCH_STR2(var)
  fmt() << "Floating-point type:" << GTBENCH_STR(GTBENCH_FLOAT);
  fmt() << "GridTools backend:" << GTBENCH_STR(GTBENCH_BACKEND);
  fmt() << "Communication backend:"
        << GTBENCH_STR(GTBENCH_COMMUNICATION_BACKEND);
#undef GTBENCH_STR
#undef GTBENCH_STR2

  // more communication setup
  auto comm_grid = communication::grid(comm_world, {n, n, nz}, num_threads);

  // analytical solution
  const real_t diffusion_coeff = 0.05;
  auto exact = verification::analytical::repeat(
      verification::analytical::advection_diffusion{diffusion_coeff},
      {(n + nz - 1) / nz, (n + nz - 1) / nz, 1});

  // benchmark executions
  std::vector<std::vector<execution::result>> all_results(num_threads);
#ifdef __CUDACC__
  auto execution_func = [&](int id = 0, int cuda_device = 0) {
    cudaSetDevice(cuda_device);
#else
  auto execution_func = [&](int id = 0) {
#endif
    for (std::size_t r = 0; r < runs; ++r) {
      all_results[id].push_back(execution::run(
          communication::sub_grid(comm_grid, id),
          numerics::advdiff_stepper(diffusion_coeff), 0.1, 1e-3, exact));
    }
  };

#if defined(GTBENCH_USE_GHEX)
  std::vector<std::thread> threads;
  threads.reserve(num_threads);
#ifdef __CUDACC__
  int cuda_device;
  cudaGetDevice(&cuda_device);
#endif
  for (int i = 0; i < num_threads; ++i) {
#ifdef __CUDACC__
    threads.push_back(std::thread{execution_func, i, cuda_device});
#else
    threads.push_back(std::thread{execution_func, i});
#endif
  }
  for (auto &t : threads)
    t.join();
  for (std::size_t i = 0; i < all_results[0].size(); ++i)
    for (std::size_t j = 1; j < all_results.size(); ++j) {
      all_results[0][i].time =
          std::max(all_results[0][i].time, all_results[j][i].time);
      all_results[0][i].error =
          std::max(all_results[0][i].error, all_results[j][i].error);
    }
#else
  execution_func();
#endif

  auto &results = all_results[0];

  // computation and reporting of median and confidence interval times
  std::sort(std::begin(results), std::end(results),
            [](execution::result const &a, execution::result const &b) {
              return a.time < b.time;
            });
  const double median_time = results[results.size() / 2].time;
  const double lower_time = results[results.size() * 25 / 1000].time;
  const double upper_time = results[results.size() * 975 / 1000].time;

  fmt() << "Median time: " << median_time << "s";
  if (runs > 100) {
    std::cout << " (95% confidence: " << lower_time << "s - " << upper_time
              << "s)";
  }

  fmt() << "Columns per second: " << (n * n / median_time);
  if (runs > 100) {
    std::cout << " (95% confidence: " << (n * n / upper_time) << " - "
              << (n * n / lower_time) << ")";
  }
  std::cout << std::endl;

  // just for safety: check if errors of all runs are (almost) the same
  for (execution::result const &r : results) {
    if (std::abs(r.error - results.front().error) > 1e-9) {
      std::cerr << "Detected error differences between the runs, "
                << "there must be something wrong!" << std::endl;
      return 1;
    }
  }

  return 0;
}
